{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "initial_id",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-26T23:39:21.084120Z",
     "start_time": "2024-08-26T23:39:20.935118Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "import re\n",
    "import os\n",
    "import numpy as np\n",
    "from pyfaidx import Fasta\n",
    "from pysam import VariantFile\n",
    "import gc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9898ccb8",
   "metadata": {},
   "source": [
    "## Generating the SNP inputs\n",
    "\n",
    "From the clair3 re_run directory (`/hpcfs/groups/phoenix-hpc-avsci/Callum_MacPhillamy/clair3_rerun/sample{01..20}.vcf.gz`) \n",
    "use the below command to collect all the VCFs for each sample.\n",
    "\n",
    "```bash\n",
    "for sample in $(seq -w 1 20)\n",
    "do\n",
    "    find . -path \"*sample${sample}/*/merge_output.vcf.gz\" | sort -V > \"sample${sample}.files\"\n",
    "done\n",
    "```\n",
    "\n",
    "This will create a file for each sample with the paths to the VCFs.\n",
    "\n",
    "We can then merge them with `bcftools` and create a single VCF for each sample.\n",
    "\n",
    "Here, `$f` is the file we created in the previous step.\n",
    "\n",
    "```bash\n",
    "for f in *.files\n",
    "do\n",
    "    bcftools concat --threads 2 -f $f -O z -o $(basename -s .files $f).vcf.gz\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1eaad95",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for '/Users/callummacphillamy/Projects/REFERENCES/ARS_UCD1.2/9913_GCA_002263795.2_current_ids.vcf.gz'\n",
      "106807756it [06:39, 267341.71it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "101997050\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Chr</th>\n",
       "      <th>MapInfo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCA12</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>2</td>\n",
       "      <td>103548215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>APAF1</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>5</td>\n",
       "      <td>63150400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARS-BFGL-BAC-10172</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>14</td>\n",
       "      <td>6371334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARS-BFGL-BAC-1020</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>14</td>\n",
       "      <td>7928189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARS-BFGL-BAC-10245</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>14</td>\n",
       "      <td>31819743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53213</th>\n",
       "      <td>UA-IFASA-9774</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>14</td>\n",
       "      <td>39201271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53214</th>\n",
       "      <td>UA-IFASA-9790</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>19</td>\n",
       "      <td>56816549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53215</th>\n",
       "      <td>UA-IFASA-9812</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>29</td>\n",
       "      <td>48012818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53216</th>\n",
       "      <td>UA-IFASA-9813</td>\n",
       "      <td>[T/C]</td>\n",
       "      <td>19</td>\n",
       "      <td>32508700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53217</th>\n",
       "      <td>UMPS</td>\n",
       "      <td>[A/G]</td>\n",
       "      <td>1</td>\n",
       "      <td>69756880</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53218 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Name    SNP Chr    MapInfo\n",
       "0                  ABCA12  [A/G]   2  103548215\n",
       "1                   APAF1  [T/C]   5   63150400\n",
       "2      ARS-BFGL-BAC-10172  [A/G]  14    6371334\n",
       "3       ARS-BFGL-BAC-1020  [T/C]  14    7928189\n",
       "4      ARS-BFGL-BAC-10245  [T/C]  14   31819743\n",
       "...                   ...    ...  ..        ...\n",
       "53213       UA-IFASA-9774  [T/C]  14   39201271\n",
       "53214       UA-IFASA-9790  [T/C]  19   56816549\n",
       "53215       UA-IFASA-9812  [T/C]  29   48012818\n",
       "53216       UA-IFASA-9813  [T/C]  19   32508700\n",
       "53217                UMPS  [A/G]   1   69756880\n",
       "\n",
       "[53218 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the Illumina manifest\n",
    "manifest = pd.read_csv('/Users/callummacphillamy/Projects/REFERENCES/UMD3.1.1/BovineSNP50_v3_A2.csv.gz',\n",
    "                       skiprows=7, header=0)\n",
    "\n",
    "# Subset to just the SNP name and RefStrand\n",
    "# subset_manifest = manifest[['Name', 'RefStrand']].copy()\n",
    "manifest.head()\n",
    "\n",
    "# Load the dbSNP vcf\n",
    "dbsnp_dict = {}\n",
    "dbsnp = VariantFile('/Users/callummacphillamy/Projects/REFERENCES/ARS_UCD1.2/9913_GCA_002263795.2_current_ids.vcf.gz')\n",
    "for record in tqdm(dbsnp.fetch()):\n",
    "    chrpos = f'{record.chrom}_{record.pos}'\n",
    "    ref = record.ref\n",
    "\n",
    "    # Skip if there are multiple alts\n",
    "    if len(record.alts) != 1:\n",
    "        continue\n",
    "    alt = record.alts[0]\n",
    "    dbsnp_dict[chrpos] = {'ref': ref, 'alt': alt}\n",
    "print(len(dbsnp_dict))\n",
    "\n",
    "dbsnp_df = pd.DataFrame.from_dict(dbsnp_dict, orient='index').reset_index()\n",
    "\n",
    "manifest_clean = manifest[['Name', 'SNP', 'Chr', 'MapInfo']].copy()\n",
    "manifest_clean = manifest_clean[~manifest_clean['Chr'].isna()]\n",
    "manifest_clean = manifest_clean[~manifest_clean['MapInfo'].isna()]\n",
    "manifest_clean['MapInfo'] = manifest_clean['MapInfo'].astype(int)\n",
    "manifest_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c90a397c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(52034, 9)\n",
      "(52034, 9)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "versa_ars = pd.read_csv('../SNP_analyses/versa50K.arsucd.bed.gz', sep='\\t',\n",
    "                          header=None,\n",
    "                          names=['chrom','pos-1','pos','name','SNP','strand'])\n",
    "\n",
    "versa_ars['ars_ChrPos'] = versa_ars['chrom'].astype(str) + '_' + versa_ars['pos'].astype(str)\n",
    "versa_ars['SNP'] = versa_ars['SNP'].str.strip('[]')\n",
    "versa_ars['ilmn_ref'] = versa_ars['SNP'].str.split('/').str[0]\n",
    "versa_ars['ilmn_alt'] = versa_ars['SNP'].str.split('/').str[1]\n",
    "print(versa_ars.shape)\n",
    "# versa_ars = versa_ars.merge(subset_manifest, left_on='name', right_on='Name', how='left')\n",
    "print(versa_ars.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "af2f72ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>ChrPos</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Pos</th>\n",
       "      <th>3DWFQ0404</th>\n",
       "      <th>3DWFQ0459</th>\n",
       "      <th>3DWFN0018</th>\n",
       "      <th>3DWFP0413</th>\n",
       "      <th>3DWFQ0177</th>\n",
       "      <th>3DWF22T0070</th>\n",
       "      <th>...</th>\n",
       "      <th>3DWF22T0549</th>\n",
       "      <th>3DWFR0281</th>\n",
       "      <th>3DWFS0069</th>\n",
       "      <th>3DWFS0202</th>\n",
       "      <th>3DWFP0398</th>\n",
       "      <th>3DWFP0416</th>\n",
       "      <th>3DWFQ0389</th>\n",
       "      <th>3DWFR0043</th>\n",
       "      <th>WGWFJ0085</th>\n",
       "      <th>3DWFR0261</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5142</th>\n",
       "      <td>BovineHD0100000005</td>\n",
       "      <td>1_16947</td>\n",
       "      <td>1</td>\n",
       "      <td>16947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5143</th>\n",
       "      <td>chr3_134833</td>\n",
       "      <td>1_19554</td>\n",
       "      <td>1</td>\n",
       "      <td>19554</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5144</th>\n",
       "      <td>BovineHD0100000033</td>\n",
       "      <td>1_111645</td>\n",
       "      <td>1</td>\n",
       "      <td>111645</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5145</th>\n",
       "      <td>BovineHD0100000035</td>\n",
       "      <td>1_120183</td>\n",
       "      <td>1</td>\n",
       "      <td>120183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5146</th>\n",
       "      <td>Hapmap43437-BTA-101873</td>\n",
       "      <td>1_135098</td>\n",
       "      <td>1</td>\n",
       "      <td>135098</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135739</th>\n",
       "      <td>BovineHD3200000384</td>\n",
       "      <td>MT_15308</td>\n",
       "      <td>MT</td>\n",
       "      <td>15308</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135740</th>\n",
       "      <td>BovineHD3200000389</td>\n",
       "      <td>MT_15505</td>\n",
       "      <td>MT</td>\n",
       "      <td>15505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135741</th>\n",
       "      <td>BovineHD3200000395</td>\n",
       "      <td>MT_15894</td>\n",
       "      <td>MT</td>\n",
       "      <td>15894</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135742</th>\n",
       "      <td>BovineHD3200000406</td>\n",
       "      <td>MT_16231</td>\n",
       "      <td>MT</td>\n",
       "      <td>16231</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135743</th>\n",
       "      <td>BovineHD3200000407</td>\n",
       "      <td>MT_16250</td>\n",
       "      <td>MT</td>\n",
       "      <td>16250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130602 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           SNP    ChrPos Chr     Pos  3DWFQ0404  3DWFQ0459  \\\n",
       "5142        BovineHD0100000005   1_16947   1   16947        NaN        NaN   \n",
       "5143               chr3_134833   1_19554   1   19554        NaN        NaN   \n",
       "5144        BovineHD0100000033  1_111645   1  111645        0.0        0.0   \n",
       "5145        BovineHD0100000035  1_120183   1  120183        NaN        NaN   \n",
       "5146    Hapmap43437-BTA-101873  1_135098   1  135098        0.0        0.0   \n",
       "...                        ...       ...  ..     ...        ...        ...   \n",
       "135739      BovineHD3200000384  MT_15308  MT   15308        2.0        2.0   \n",
       "135740      BovineHD3200000389  MT_15505  MT   15505        NaN        NaN   \n",
       "135741      BovineHD3200000395  MT_15894  MT   15894        NaN        NaN   \n",
       "135742      BovineHD3200000406  MT_16231  MT   16231        2.0        2.0   \n",
       "135743      BovineHD3200000407  MT_16250  MT   16250        NaN        NaN   \n",
       "\n",
       "        3DWFN0018  3DWFP0413  3DWFQ0177  3DWF22T0070  ...  3DWF22T0549  \\\n",
       "5142          NaN        NaN        NaN          NaN  ...          NaN   \n",
       "5143          NaN        NaN        NaN          NaN  ...          NaN   \n",
       "5144          0.0        0.0        0.0          0.0  ...          0.0   \n",
       "5145          NaN        NaN        NaN          NaN  ...          NaN   \n",
       "5146          0.0        0.0        0.0          0.0  ...          0.0   \n",
       "...           ...        ...        ...          ...  ...          ...   \n",
       "135739        2.0        2.0        2.0          2.0  ...          2.0   \n",
       "135740        NaN        NaN        NaN          NaN  ...          NaN   \n",
       "135741        NaN        NaN        NaN          NaN  ...          NaN   \n",
       "135742        2.0        2.0        2.0          2.0  ...          2.0   \n",
       "135743        NaN        NaN        NaN          NaN  ...          NaN   \n",
       "\n",
       "        3DWFR0281  3DWFS0069  3DWFS0202  3DWFP0398  3DWFP0416  3DWFQ0389  \\\n",
       "5142          NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "5143          NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "5144          NaN        0.0        0.0        0.0        0.0        0.0   \n",
       "5145          NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "5146          0.0        0.0        0.0        0.0        0.0        0.0   \n",
       "...           ...        ...        ...        ...        ...        ...   \n",
       "135739        2.0        2.0        2.0        2.0        2.0        2.0   \n",
       "135740        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "135741        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "135742        2.0        2.0        2.0        2.0        2.0        2.0   \n",
       "135743        NaN        NaN        NaN        NaN        NaN        NaN   \n",
       "\n",
       "        3DWFR0043  WGWFJ0085  3DWFR0261  \n",
       "5142          NaN        NaN        NaN  \n",
       "5143          NaN        NaN        NaN  \n",
       "5144          0.0        0.0        0.0  \n",
       "5145          NaN        NaN        NaN  \n",
       "5146          0.0        0.0        0.0  \n",
       "...           ...        ...        ...  \n",
       "135739        2.0        2.0        2.0  \n",
       "135740        NaN        NaN        NaN  \n",
       "135741        NaN        NaN        NaN  \n",
       "135742        2.0        2.0        2.0  \n",
       "135743        NaN        NaN        NaN  \n",
       "\n",
       "[130602 rows x 29 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the original SNP dataframe. I.e. Joe's genotypes on Versa50K\n",
    "og_snps = pd.read_parquet('../SNP_analyses/genotypes/3DG_Nucleus_array_genotypes.parquet')\n",
    "og_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7647901b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ars_ChrPos</th>\n",
       "      <th>3DWFQ0404</th>\n",
       "      <th>3DWFQ0459</th>\n",
       "      <th>3DWFN0018</th>\n",
       "      <th>3DWFP0413</th>\n",
       "      <th>3DWFQ0177</th>\n",
       "      <th>3DWF22T0070</th>\n",
       "      <th>3DWFP0355</th>\n",
       "      <th>WGWFL0149</th>\n",
       "      <th>3DWFS0137</th>\n",
       "      <th>...</th>\n",
       "      <th>3DWF22T0549</th>\n",
       "      <th>3DWFR0281</th>\n",
       "      <th>3DWFS0069</th>\n",
       "      <th>3DWFS0202</th>\n",
       "      <th>3DWFP0398</th>\n",
       "      <th>3DWFP0416</th>\n",
       "      <th>3DWFQ0389</th>\n",
       "      <th>3DWFR0043</th>\n",
       "      <th>WGWFJ0085</th>\n",
       "      <th>3DWFR0261</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_752778</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>--</td>\n",
       "      <td>AA</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>--</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_776231</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>...</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_907810</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>...</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_1032564</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_1110393</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AG</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>...</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50042</th>\n",
       "      <td>Y_23289148</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50043</th>\n",
       "      <td>Y_24113061</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50044</th>\n",
       "      <td>Y_24881352</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50045</th>\n",
       "      <td>Y_25004275</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50046</th>\n",
       "      <td>Y_58123428</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>...</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "      <td>--</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50047 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ars_ChrPos 3DWFQ0404 3DWFQ0459 3DWFN0018 3DWFP0413 3DWFQ0177  \\\n",
       "0        1_752778        AA        AA        AA        AA        AA   \n",
       "1        1_776231        TT        TT        TT        TT        TT   \n",
       "2        1_907810        CC        CC        CC        CC        CC   \n",
       "3       1_1032564        GG        GG        GG        GG        GG   \n",
       "4       1_1110393        AA        AG        AG        AA        AA   \n",
       "...           ...       ...       ...       ...       ...       ...   \n",
       "50042  Y_23289148        --        --        --        --        --   \n",
       "50043  Y_24113061        --        --        --        --        --   \n",
       "50044  Y_24881352        --        --        --        --        --   \n",
       "50045  Y_25004275        --        --        --        --        --   \n",
       "50046  Y_58123428        --        --        --        --        --   \n",
       "\n",
       "      3DWF22T0070 3DWFP0355 WGWFL0149 3DWFS0137  ... 3DWF22T0549 3DWFR0281  \\\n",
       "0              AA        AA        --        AA  ...          AA        --   \n",
       "1              TT        TT        TT        TT  ...          TT        TT   \n",
       "2              CC        CC        CC        CC  ...          CC        CC   \n",
       "3              GG        GG        GG        GG  ...          GG        GG   \n",
       "4              AA        AG        AA        AA  ...          AA        AA   \n",
       "...           ...       ...       ...       ...  ...         ...       ...   \n",
       "50042          --        --        --        --  ...          --        --   \n",
       "50043          --        --        --        --  ...          --        --   \n",
       "50044          --        --        --        --  ...          --        --   \n",
       "50045          --        --        --        --  ...          --        --   \n",
       "50046          --        --        --        --  ...          --        --   \n",
       "\n",
       "      3DWFS0069 3DWFS0202 3DWFP0398 3DWFP0416 3DWFQ0389 3DWFR0043 WGWFJ0085  \\\n",
       "0            AA        AA        AA        AA        AA        AA        AA   \n",
       "1            TT        TT        TT        TT        TT        TT        TT   \n",
       "2            CC        CC        CC        CC        CC        CC        CC   \n",
       "3            GG        GG        GG        GG        GG        GG        GG   \n",
       "4            AA        AA        AA        AA        AA        AA        AA   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "50042        --        --        --        --        --        --        --   \n",
       "50043        --        --        --        --        --        --        --   \n",
       "50044        --        --        --        --        --        --        --   \n",
       "50045        --        --        --        --        --        --        --   \n",
       "50046        --        --        --        --        --        --        --   \n",
       "\n",
       "      3DWFR0261  \n",
       "0            AA  \n",
       "1            TT  \n",
       "2            CC  \n",
       "3            GG  \n",
       "4            AA  \n",
       "...         ...  \n",
       "50042        --  \n",
       "50043        --  \n",
       "50044        --  \n",
       "50045        --  \n",
       "50046        --  \n",
       "\n",
       "[50047 rows x 26 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "versa_ars_umd_genotypes = pd.merge(versa_ars[['name','ars_ChrPos','ilmn_ref','ilmn_alt']], og_snps, left_on='name',\n",
    "                               right_on='SNP',\n",
    "                               how='left')\n",
    "versa_ars_umd_genotypes\n",
    "\n",
    "# Map the floats back to the actual alleles\n",
    "float_to_allele_dict = {}\n",
    "for row in versa_ars_umd_genotypes.itertuples():\n",
    "    chrpos = row.ars_ChrPos\n",
    "    float_to_allele_dict[chrpos] = {0. : f'{row.ilmn_ref}{row.ilmn_ref}',\n",
    "                                    1. : f'{row.ilmn_ref}{row.ilmn_alt}',\n",
    "                                    2. : f'{row.ilmn_alt}{row.ilmn_alt}',\n",
    "                                    np.nan: '--'}\n",
    "\n",
    "ars_genotype_dict = {}\n",
    "for sample in versa_ars_umd_genotypes.columns[8:]:\n",
    "    ars_genotype_dict[sample] = {}\n",
    "    for row in versa_ars_umd_genotypes.loc[:, ['ars_ChrPos', sample]].itertuples():\n",
    "        chrpos = row.ars_ChrPos\n",
    "        ars_genotype_dict[sample][chrpos] = float_to_allele_dict[chrpos].get(row[2], '--')\n",
    "        \n",
    "\n",
    "ars_genotype_df = pd.DataFrame.from_dict(ars_genotype_dict, orient='index').T\n",
    "ars_genotype_df = ars_genotype_df.rename_axis('ars_ChrPos').reset_index()\n",
    "ars_genotype_df\n",
    "\n",
    "chrpos_set = set(versa_ars_umd_genotypes['ars_ChrPos'].values)\n",
    "ars_genotype_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c024ce1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ChrPos</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>sample</th>\n",
       "      <th>genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_25784</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_25787</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0/1</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_25797</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0/1</td>\n",
       "      <td>CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_25810</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0/1</td>\n",
       "      <td>CG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_25828</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0/1</td>\n",
       "      <td>CA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9019576</th>\n",
       "      <td>Y_59471086</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0/1</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9019577</th>\n",
       "      <td>Y_59471104</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0/1</td>\n",
       "      <td>CG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9019578</th>\n",
       "      <td>Y_59471116</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0/1</td>\n",
       "      <td>CG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9019588</th>\n",
       "      <td>Y_59474743</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0/1</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9019589</th>\n",
       "      <td>Y_59474752</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1/1</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7497275 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ChrPos REF ALT sample genotype\n",
       "0           1_25784   A   T    0/1       AT\n",
       "1           1_25787   A   G    0/1       AG\n",
       "2           1_25797   C   T    0/1       CT\n",
       "3           1_25810   C   G    0/1       CG\n",
       "4           1_25828   C   A    0/1       CA\n",
       "...             ...  ..  ..    ...      ...\n",
       "9019576  Y_59471086   A   G    0/1       AG\n",
       "9019577  Y_59471104   C   G    0/1       CG\n",
       "9019578  Y_59471116   C   G    0/1       CG\n",
       "9019588  Y_59474743   G   A    0/1       GA\n",
       "9019589  Y_59474752   T   G    1/1       GG\n",
       "\n",
       "[7497275 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_dict = {\n",
    "    'sample01':'3DWFQ0404',\n",
    "    'sample02':'3DWFQ0459',\n",
    "    'sample03':'3DWFN0018',\n",
    "    'sample04':'3DWFP0413',\n",
    "    'sample05':'3DWFQ0177',\n",
    "    'sample06':'3DWF22T0070',\n",
    "    'sample07':'3DWFP0355',\n",
    "    'sample08':'WGWFL0149',\n",
    "    'sample09':'3DWFS0137',\n",
    "    'sample10':'3DWFR0017',\n",
    "    'sample11':'3DWF22T0415',\n",
    "    'sample12':'3DWF22T0240',\n",
    "    'sample13':'3DWFQ0458',\n",
    "    'sample14':'3DWFS0123',\n",
    "    'sample15':'3DWF22T0036',\n",
    "    'sample16':'3DWF22T0549',\n",
    "    'sample17':'3DWFR0281',\n",
    "    'sample18':'3DWFS0069',\n",
    "    'sample19':'3DWFS0202',\n",
    "    'sample20':'3DWFP0398'\n",
    "}\n",
    "\n",
    "sample = pd.read_csv(\n",
    "    '/Users/callummacphillamy/Projects/tuwa_manuscript/TuWa_manuscript/SNP_analyses/genotypes/clair3_rerun/sample01.vcf.gz',\n",
    "    sep='\\t', comment='#', header=None, dtype={0:str},\n",
    "    names=['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','sample'])\n",
    "\n",
    "clean_sample = sample[sample['FILTER'] != 'LowQual']\n",
    "\n",
    "# Remove indels/ Only consider SNPs\n",
    "clean_sample = clean_sample[clean_sample['REF'].str.len() == 1]\n",
    "clean_sample = clean_sample[clean_sample['ALT'].str.len() == 1]\n",
    "clean_sample = clean_sample.copy()\n",
    "\n",
    "\n",
    "clean_sample['ChrPos'] = clean_sample['CHROM'].astype(str) + '_' + clean_sample['POS'].astype(str)\n",
    "clean_sample = clean_sample[['ChrPos','REF','ALT','sample']]\n",
    "clean_sample.copy()\n",
    "\n",
    "clean_sample['sample'] = clean_sample['sample'].str.split(':').str[0]\n",
    "\n",
    "# Turn the sample column into the alleles genotype. E.g. 0/1 -> AT\n",
    "def get_genotype(row):\n",
    "    genotype = row['sample']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    if genotype == '0/0':\n",
    "        return ref + ref\n",
    "    elif genotype == '0/1':\n",
    "        return ref + alt\n",
    "    elif genotype == '1/1':\n",
    "        return alt + alt\n",
    "    else:\n",
    "        return '--'\n",
    "\n",
    "# Apply the function to the dataframe\n",
    "clean_sample['genotype'] = clean_sample.apply(get_genotype, axis=1)\n",
    "\n",
    "clean_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f132e55e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ars_ChrPos</th>\n",
       "      <th>3DWFQ0404</th>\n",
       "      <th>ChrPos</th>\n",
       "      <th>genotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1110393</td>\n",
       "      <td>AA</td>\n",
       "      <td>1_1110393</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_1291908</td>\n",
       "      <td>CC</td>\n",
       "      <td>1_1291908</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_1350801</td>\n",
       "      <td>AG</td>\n",
       "      <td>1_1350801</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_1566539</td>\n",
       "      <td>CC</td>\n",
       "      <td>1_1566539</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_1604940</td>\n",
       "      <td>CC</td>\n",
       "      <td>1_1604940</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26755</th>\n",
       "      <td>Y_490228</td>\n",
       "      <td>AG</td>\n",
       "      <td>Y_490228</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26756</th>\n",
       "      <td>Y_549318</td>\n",
       "      <td>AG</td>\n",
       "      <td>Y_549318</td>\n",
       "      <td>CT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26757</th>\n",
       "      <td>Y_614721</td>\n",
       "      <td>AG</td>\n",
       "      <td>Y_614721</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26758</th>\n",
       "      <td>Y_780515</td>\n",
       "      <td>TC</td>\n",
       "      <td>Y_780515</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26759</th>\n",
       "      <td>Y_986552</td>\n",
       "      <td>TC</td>\n",
       "      <td>Y_986552</td>\n",
       "      <td>GA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26760 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ars_ChrPos 3DWFQ0404     ChrPos genotype\n",
       "0      1_1110393        AA  1_1110393       TT\n",
       "1      1_1291908        CC  1_1291908       CC\n",
       "2      1_1350801        AG  1_1350801       GA\n",
       "3      1_1566539        CC  1_1566539       GG\n",
       "4      1_1604940        CC  1_1604940       CC\n",
       "...          ...       ...        ...      ...\n",
       "26755   Y_490228        AG   Y_490228       AG\n",
       "26756   Y_549318        AG   Y_549318       CT\n",
       "26757   Y_614721        AG   Y_614721       AG\n",
       "26758   Y_780515        TC   Y_780515       AG\n",
       "26759   Y_986552        TC   Y_986552       GA\n",
       "\n",
       "[26760 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We will use the DBSNP to resolve what the ref and alt alleles should be\n",
    "sample_to_compare = sample_dict['sample01']\n",
    "ars_genotype_df[['ars_ChrPos',sample_to_compare]]\n",
    "comparison = pd.merge(ars_genotype_df[['ars_ChrPos',sample_to_compare]], clean_sample[['ChrPos','genotype']], left_on='ars_ChrPos',\n",
    "                       right_on='ChrPos',\n",
    "                       how='inner')\n",
    "\n",
    "comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4b8ff9b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "26760it [00:00, 311032.71it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Correct: 25764\n",
      "Incorrect: 687\n",
      "Correct: 0.9740274469774299\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "big_comparison = comparison.merge(dbsnp_df, left_on='ChrPos', right_on='index', how='left')\n",
    "big_comparison\n",
    "\n",
    "# Compare the genotypes from the two samples, relative to the dbSNP ref and alt\n",
    "correct = 0\n",
    "incorrect = []\n",
    "\n",
    "complementary = {'AA':'TT',\n",
    "                 'CC':'GG',\n",
    "                 'TT':'AA',\n",
    "                 'GG':'CC',\n",
    "                 'AC':'TG',\n",
    "                 'TG':'AC',\n",
    "                 'TC':'AG',\n",
    "                 'AG':'TC'}\n",
    "\n",
    "\n",
    "missing = 0\n",
    "\n",
    "for row in tqdm(big_comparison.itertuples()):\n",
    "\n",
    "    # Skip if there are any NaNs in the row\n",
    "    if sum([pd.isna(val) for val in row]) > 0:\n",
    "        continue\n",
    "        \n",
    "    true_ref = row.ref\n",
    "    true_alt = row.alt\n",
    "\n",
    "    true_hom_ref = true_ref + true_ref\n",
    "    true_het = true_ref + true_alt\n",
    "    true_het_other_way = true_alt + true_ref\n",
    "    true_hom_alt = true_alt + true_alt\n",
    "\n",
    "    heterozygous = [true_het, true_het_other_way]\n",
    "    \n",
    "\n",
    "    ont_gt = row.genotype\n",
    "    array_gt = row[2]\n",
    "\n",
    "    if ont_gt == '--' or array_gt == '--':\n",
    "        missing += 1\n",
    "        continue\n",
    "\n",
    "    # Case 1: Both genotypes match the true homozygous ref\n",
    "    if ont_gt == true_hom_ref and array_gt == true_hom_ref:\n",
    "        correct += 1\n",
    "    \n",
    "    # Case 2: Both genotypes match the true heterozygous\n",
    "    elif ont_gt in heterozygous and array_gt in heterozygous:\n",
    "        correct += 1\n",
    "\n",
    "    # Case 3: Both genotypes match the true homozygous alt\n",
    "    elif ont_gt == true_hom_alt and array_gt == true_hom_alt:\n",
    "        correct += 1\n",
    "\n",
    "    # Case 4: Array complement matches the true homozygous ref\n",
    "    elif complementary.get(array_gt, '') == true_hom_ref and ont_gt == true_hom_ref:\n",
    "        correct += 1\n",
    "    \n",
    "    # Case 5: Array complement matches the true homozygous alt\n",
    "    elif complementary.get(array_gt, '') in true_hom_alt and ont_gt in true_hom_alt:\n",
    "        correct += 1\n",
    "    \n",
    "    # Case 6: Array is heterozygous and complementary to the true heterozygous\n",
    "    elif complementary.get(array_gt, '') in heterozygous and ont_gt in heterozygous:\n",
    "        correct += 1\n",
    "    \n",
    "    # Case 7: Mismatch\n",
    "    else:\n",
    "        incorrect.append(row)\n",
    "    \n",
    "    \n",
    "print(f'Correct: {correct}')\n",
    "print(f'Incorrect: {len(incorrect)}')\n",
    "print(f'Correct: {correct/(correct + len(incorrect))}')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc5fdf34",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(incorrect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a589aaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_to_compare = sample_dict['sample01']\n",
    "comparison = pd.merge(ars_genotype_df[['ars_ChrPos',sample_to_compare]], clean_sample[['ChrPos','genotype']], left_on='ars_ChrPos',\n",
    "                       right_on='ChrPos',\n",
    "                       how='inner')\n",
    "comparison\n",
    "\n",
    "# correct if complement means TC and AG/GA are considered the same, for example.\n",
    "\n",
    "complement_dict = {'A':'T',\n",
    "                   'T':'A',\n",
    "                   'C':'G',\n",
    "                   'G':'C'}\n",
    "\n",
    "def complement(seq):\n",
    "    \"\"\"\n",
    "    Returns the complement of a DNA sequence.\n",
    "    \"\"\"\n",
    "    return ''.join(complement_dict.get(base, base) for base in seq)\n",
    "\n",
    "\n",
    "result = {'correct_as_is':0,\n",
    "          'correct_if_complement':0,\n",
    "          'incorrect':[]}\n",
    "for row in tqdm(comparison.itertuples()):\n",
    "    ont = row.genotype\n",
    "    ars = row[2]\n",
    "\n",
    "    ont = ''.join(sorted(ont))\n",
    "    ars = ''.join(sorted(ars))\n",
    "\n",
    "    if '--' in [ont, ars]:\n",
    "        continue\n",
    "\n",
    "    if ont == ars:\n",
    "        result['correct_as_is'] += 1\n",
    "    elif ont == complement(ars):\n",
    "        result['correct_if_complement'] += 1\n",
    "    else:\n",
    "        result['incorrect'].append([row.ars_ChrPos, ont, ars])\n",
    "print(result['correct_as_is'])\n",
    "print(result['correct_if_complement'])\n",
    "print(len(result['incorrect']))\n",
    "print('Total potentially correct: ', result['correct_as_is'] + result['correct_if_complement'])\n",
    "print('Total examined: ', result['correct_as_is'] + result['correct_if_complement'] + len(result['incorrect']))\n",
    "print('Percentage correct: ', (result['correct_as_is'] + result['correct_if_complement']) / (result['correct_as_is'] + result['correct_if_complement'] + len(result['incorrect'])) * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0250be38",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genotype(gt_float, ref, alt, comp_dict, use_comp=False):\n",
    "    if gt_float == 0.:\n",
    "        if use_comp:\n",
    "            return comp_dict[ref]+comp_dict[ref]\n",
    "        else:\n",
    "            return ref+ref\n",
    "    elif gt_float == 1.:\n",
    "        if use_comp:\n",
    "            return comp_dict[ref]+comp_dict[alt]\n",
    "        else:\n",
    "            return ref+alt\n",
    "    elif gt_float == 2.:\n",
    "        if use_comp:\n",
    "            return comp_dict[alt]+comp_dict[alt]\n",
    "        else:\n",
    "            return alt+alt\n",
    "    else:\n",
    "        return np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375c9ce7",
   "metadata": {},
   "outputs": [],
   "source": [
    "correct = 0\n",
    "half_correct = 0\n",
    "incorrect = 0\n",
    "ont_missing = 0\n",
    "ilmn_missing = 0\n",
    "valid = 0\n",
    "for row in tqdm(comparison.itertuples()):\n",
    "    if len(row.REF) > 1 or len(row.ALT) > 1:\n",
    "        continue\n",
    "    if row.sample not in [1.,2.,0.] and row._9 not in [1.,2.,0.]:\n",
    "        ont_missing += 1\n",
    "        ilmn_missing += 1\n",
    "        continue\n",
    "    elif row.sample not in [1.,2.,0.]:\n",
    "        ont_missing += 1\n",
    "        continue\n",
    "    elif row._9 not in [1.,2.,0.]:\n",
    "        ilmn_missing += 1\n",
    "        continue\n",
    "    elif row.ilmn_ref in ['I','D']:\n",
    "        continue\n",
    "    \n",
    "    # print(row)\n",
    "    ont_alleles = [row.REF, row.ALT]\n",
    "    ilmn_alleles = [row.ilmn_ref, row.ilmn_alt]\n",
    "    ont_gt = get_genotype(row.sample, row.REF, row.ALT, comp_dict=comp_dict, use_comp=False)\n",
    "    comp_ont_gt = get_genotype(gt_float = row.sample,\n",
    "                               ref = row.REF,\n",
    "                               alt = row.ALT, comp_dict=comp_dict, use_comp=True)\n",
    "    ilmn_gt = get_genotype(row._9, row.ilmn_ref, row.ilmn_alt, comp_dict=comp_dict, use_comp=False)\n",
    "\n",
    "    if ont_gt is np.nan or ilmn_gt is np.nan:\n",
    "        print(row)\n",
    "        raise ValueError('genotype is nan')\n",
    "\n",
    "    if ont_gt == ilmn_gt:\n",
    "        correct += 1\n",
    "    \n",
    "    elif ont_gt[0] in ilmn_gt or ont_gt[1] in ilmn_gt:\n",
    "        half_correct += 1\n",
    "    \n",
    "    elif ont_gt != ilmn_gt:\n",
    "\n",
    "        # Try to correct the genotype\n",
    "        if comp_ont_gt == ilmn_gt:\n",
    "            correct += 1\n",
    "        elif comp_ont_gt[0] in ilmn_gt or comp_ont_gt[1] in ilmn_gt:\n",
    "            half_correct += 1\n",
    "        else:\n",
    "            incorrect += 1\n",
    "    \n",
    "    valid += 1\n",
    "\n",
    "print('Of {} valid SNPs:'.format(valid))\n",
    "print(correct, np.around(correct / valid, decimals=2), 'proportion were concordant')\n",
    "print(half_correct, np.around(half_correct / valid, decimals=2), 'proportion were half concordant')\n",
    "print(incorrect, np.around(incorrect / valid,decimals=2), 'proportion were discordant')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fc64d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for row in comparison.itertuples():\n",
    "    if row.ALT == '.':\n",
    "        if row.REF == row.ilmn_ref:\n",
    "            if row.sample == row._8:\n",
    "                correct += 1\n",
    "            elif np.abs(row.sample - row._8) == 1:\n",
    "                half_correct += 1\n",
    "            else:\n",
    "                incorrect += 1\n",
    "    \n",
    "    # REF and ALT match Illumina REF and ALT\n",
    "    if row.REF == row.ilmn_ref and row.ALT == row.ilmn_alt:\n",
    "        if row.sample == row._8:\n",
    "            correct += 1\n",
    "        elif np.abs(row.sample - row._8) == 1:\n",
    "            half_correct += 1\n",
    "        else:\n",
    "            incorrect += 1\n",
    "    \n",
    "    # REF and ALT don't match Illumina REF and ALT\n",
    "    elif row.REF != row.ilmn_ref or row.ALT != row.ilmn_alt:\n",
    "        \n",
    "        # See if using the complement fixes it.\n",
    "        if row.REF == comp_dict[row.ilmn_ref] and row.ALT == comp_dict[row.ilmn_alt]:\n",
    "            if row.sample == row._8:\n",
    "                correct += 1\n",
    "            elif np.abs(row.sample - row._8) == 1:\n",
    "                half_correct += 1\n",
    "            else:\n",
    "                incorrect += 1\n",
    "        \n",
    "        \n",
    "        if row.sample == row._8:\n",
    "            incorrect += 1\n",
    "        elif np.abs(row.sample - row._8) == 1:\n",
    "            half_correct += 1\n",
    "        else:\n",
    "            correct += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fb1087277623590",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-21T03:12:00.214513Z",
     "start_time": "2024-08-21T03:11:59.191716Z"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "#for i in {1..29}\n",
    "#do\n",
    "#    grep -w \"chr${i}\" /Users/callummacphillamy/Projects/bovine_pangenome/TuWa_manuscript/SNP_analyses/SNP_bed_files/versa50K_SNPs.liftover.bed > /Users/callummacphillamy/Projects/bovine_pangenome/TuWa_manuscript/SNP_analyses/versa50K_SNPs.chr${i}.bed\n",
    "#done\n",
    "#grep \"chrX\" /Users/callummacphillamy/Projects/bovine_pangenome/TuWa_manuscript/SNP_analyses/SNP_bed_files/versa50K_SNPs.liftover.bed > /Users/callummacphillamy/Projects/bovine_pangenome/TuWa_manuscript/SNP_analyses/versa50K_SNPs.chrX.bed\n",
    "#grep \"chrY\" /Users/callummacphillamy/Projects/bovine_pangenome/TuWa_manuscript/SNP_analyses/SNP_bed_files/versa50K_SNPs.liftover.bed > /Users/callummacphillamy/Projects/bovine_pangenome/TuWa_manuscript/SNP_analyses/versa50K_SNPs.chrY.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f84c3cdf3937f2a2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:14:41.983359Z",
     "start_time": "2024-08-26T23:57:37.252969Z"
    }
   },
   "outputs": [],
   "source": [
    "#`~/software/gatk-4.4.0.0/gatk GenotypeGVCFs -R ~/Projects/REFERENCES/bosTau6/bosTau6.fa -V merge_output.gvcf.gz -O test.vcf.gz -L test.bed -all-sites`\n",
    "# files = glob.glob('../RESULTS/mapped_to_bosTau6/SNPs/*/*/*.gvcf.gz')\n",
    "files = glob.glob('../RESULTS/mapped_to_bosTau6/SNPs/sample06/*/*.gvcf.gz')\n",
    "regions = '../SNP_analyses/SNP_bed_files'\n",
    "for file in files:\n",
    "    sample = file.split('/')[-3]\n",
    "    \n",
    "    if sample not in file:\n",
    "        continue\n",
    "    chromosome = file.split('/')[-2]\n",
    "    region = os.path.join(regions, f'versa50K_SNPs.{chromosome}.bed')\n",
    "    vcf = os.path.join(\"../SNP_analyses/genotypes/mapped_to_bosTau6\",\n",
    "                       f\"{sample}.{chromosome}.versa50K.vcf.gz\")\n",
    "    ! ~/software/gatk-4.4.0.0/gatk GenotypeGVCFs -R ~/Projects/REFERENCES/bosTau6/bosTau6_w_wagyuY.fa -V {file} -O {vcf} -L {region} -all-sites\n",
    "    \n",
    "## There seems to be some issue with the length of chromosome 3 in the reference genome, vs what's in the bed file...\n",
    "## I think the best solution will be to drop those SNPs and hope it's not too many..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e7c4446ee632701",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:17:20.555402Z",
     "start_time": "2024-08-27T00:17:20.221158Z"
    }
   },
   "outputs": [],
   "source": [
    "# samples = ['01','03','05','12']\n",
    "samples = ['06']\n",
    "for sample in samples:\n",
    "    ! bcftools concat -O z ../SNP_analyses/genotypes/mapped_to_bosTau6/sample{sample}.chr*.vcf.gz | bcftools sort -O z > ../SNP_analyses/genotypes/mapped_to_bosTau6/sample{sample}.versa50K.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffb61b6cae1adf73",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:32:29.548544Z",
     "start_time": "2024-08-27T00:28:51.746025Z"
    }
   },
   "outputs": [],
   "source": [
    "DBSNP = '/Users/callummacphillamy/Projects/REFERENCES/bosTau6/Bos_taurus.vcf'\n",
    "dbsnp = pd.read_csv(DBSNP, sep=\"\\t\", comment=\"#\", dtype={0:str}, header=None)\n",
    "dbsnp.columns = ['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO']\n",
    "dbsnp = dbsnp.copy()\n",
    "dbsnp = dbsnp[dbsnp['REF'].isin(['A','C','G','T'])]\n",
    "\n",
    "dbsnp_dict_df = {'ChrPos':[],\n",
    "         'Ref':[],\n",
    "         'Alt':[],\n",
    "         'ID':[]}\n",
    "dbsnp_dict = {}\n",
    "for row in tqdm(dbsnp.itertuples()):\n",
    "    #print(row)\n",
    "    info = row.INFO.split(';')\n",
    "    type = info[1].split('=')[1]\n",
    "    if type != 'SNV':\n",
    "        continue\n",
    "    chrpos = 'chr' + str(row.CHROM) + '_' + str(row.POS)\n",
    "    alt = row.REF\n",
    "    ref = row.ALT\n",
    "    id = row.ID\n",
    "    dbsnp_dict_df['ChrPos'].append(chrpos)\n",
    "    dbsnp_dict_df['Ref'].append(ref)\n",
    "    dbsnp_dict_df['Alt'].append(alt)\n",
    "    dbsnp_dict_df['ID'].append(id)\n",
    "    dbsnp_dict[chrpos] = {'Ref':ref,\n",
    "                            'Alt':alt,\n",
    "                            'ID':id}\n",
    "del dbsnp\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ac33010a2a6f774",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:40:42.562179Z",
     "start_time": "2024-08-27T00:40:42.547770Z"
    }
   },
   "outputs": [],
   "source": [
    "def load_vcf(vcf, sample):\n",
    "    vcf = pd.read_csv(vcf, sep='\\t', comment='#', header=None)\n",
    "    vcf.columns = ['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT',sample]\n",
    "    return vcf\n",
    "\n",
    "unambiguous_snps = {'AC': ('A', 'C', 'TOP'),\n",
    "                        'AG': ('A', 'G', 'TOP'),\n",
    "                        'TC': ('T', 'C', 'BOT'),\n",
    "                        'TG': ('T', 'G', 'BOT'),\n",
    "                        'CA': ('A', 'C', 'TOP'),\n",
    "                        'GA': ('A', 'G', 'TOP'),\n",
    "                        'CT': ('T', 'C', 'BOT'),\n",
    "                        'GT': ('T', 'G', 'BOT')}\n",
    "\n",
    "def get_sequence(reference=None, snp_pos=None, ref_allele=None):\n",
    "    \"\"\"\n",
    "    Get the sequence around a SNP position.\n",
    "    Args:\n",
    "        reference: A pyfaidx Fasta instance\n",
    "        snp_pos: a tuple with form (chr, pos)\n",
    "        ref_allele: the ref allele from the VCF file\n",
    "\n",
    "    Returns:\n",
    "        sequence: str, the sequence around the SNP position\n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    complementary_bases = {'A':'T',\n",
    "                           'T':'A',\n",
    "                           'C':'G',\n",
    "                           'G':'C'}\n",
    "\n",
    "    if reference is None:\n",
    "        raise ValueError('Reference should be a pyfaidx.Fasta instance')\n",
    "    if snp_pos is None:\n",
    "        raise ValueError('snp_pos should be a tuple with form (chr, pos)')\n",
    "\n",
    "\n",
    "    # Check nucleotide at chr pos matches the ref allele\n",
    "    ref_position = reference[str(snp_pos[0])][snp_pos[1] - 1: snp_pos[1]]\n",
    "    print(ref_position)\n",
    "\n",
    "    if ref_allele.upper() != str(ref_position).upper() and ref_allele.upper() != complementary_bases[str(ref_position).upper()]:\n",
    "        print(f'Ref allele: {ref_allele}, Allele at position: {ref_position}')\n",
    "        raise ValueError('The nucleotide at the provided position does not match'\n",
    "                         ' the reference allele')\n",
    "\n",
    "\n",
    "    sequence = str(reference[str(snp_pos[0])][snp_pos[1] - 41 : snp_pos[1] + 40])\n",
    "    if sequence[40:41].upper() != ref_allele.upper() and sequence[40:41].upper() != complementary_bases[ref_allele.upper()]:\n",
    "        raise ValueError('Midpoint of sequence doesn\\'t match the ref allele')\n",
    "\n",
    "    return sequence.upper()\n",
    "\n",
    "def resolve_ambiguous_snp(seq=None, ref_allele=None, alt_allele=None):\n",
    "    \"\"\"\n",
    "    Resolve ambiguous SNPs (A/T and C/G) using the sequence walk method.\n",
    "\n",
    "    Args:\n",
    "        seq: str, sequence around the SNP. The SNP should be in the center of \n",
    "            the sequence with an equal number of nucleotides up and down stream.\n",
    "        ref_allele: str, The reference allele from the VCF file\n",
    "        alt_allele: str, the alternate allele from the VCF file\n",
    "\n",
    "    Returns:\n",
    "        allele_a, allele_b, strand - The reference allele, alternate allele and strand\n",
    "\n",
    "    Examples:\n",
    "        >>> resolve_ambiguous_snp(seq=\"ACGGGGACAGATATGTTAACT\", ref_allele='A', alt_allele='T')\n",
    "        ('T', 'A', 'BOT')\n",
    "\n",
    "    Raises:\n",
    "        ValueError: If the ref allele does not match the nucleotide at the midpoint of the sequence\n",
    "        ValueError: If the strand of the SNP cannot be resolved\n",
    "    \"\"\"\n",
    "\n",
    "    _unambiguous_snps = {'AC': None,\n",
    "                         'AG': None,\n",
    "                         'TC': None,\n",
    "                         'TG': None,\n",
    "                         'CA': None,\n",
    "                         'GA': None,\n",
    "                         'CT': None,\n",
    "                         'GT': None}\n",
    "\n",
    "    # complementary_bases = {'A':'T',\n",
    "    #                        'T':'A',\n",
    "    #                        'C':'G',\n",
    "    #                        'G':'C'}\n",
    "\n",
    "    if len(seq) % 2 == 0:\n",
    "        raise ValueError(\"Sequence length is even. This will cause issues.\")\n",
    "\n",
    "    seq_midpoint = len(seq) // 2\n",
    "\n",
    "    if seq[seq_midpoint] != ref_allele and seq[seq_midpoint] != alt_allele: #and seq[seq_midpoint] != complementary_bases[ref_allele]:\n",
    "        print('Neither alleles match the midpoint')\n",
    "        print(f\"Ref allele: {ref_allele}, Allele at midpoint: {seq[seq_midpoint]}\")\n",
    "        raise ValueError(\"Ref allele does not match the expected nucleotide at midpoint\")\n",
    "\n",
    "    # Begin sequence walk to resolve strand and A B alleles\n",
    "    walk_toward_5_prime = [i for i in range(seq_midpoint - 1, 0-1, -1)]\n",
    "    walk_toward_3_prime = [i for i in range(seq_midpoint + 1, len(seq))]\n",
    "\n",
    "    walking_nt_positions = zip(walk_toward_5_prime, walk_toward_3_prime)\n",
    "\n",
    "    if ref_allele is None:\n",
    "        raise TypeError(\"ref_allele should not be None\")\n",
    "\n",
    "    # if alt_allele is None:\n",
    "    #     alt_allele = ref_allele\n",
    "\n",
    "    for i, j in walking_nt_positions:\n",
    "        _5_prime_nt = seq[i]\n",
    "        _3_prime_nt = seq[j]\n",
    "        _pairing = _5_prime_nt + _3_prime_nt\n",
    "        if _pairing in _unambiguous_snps.keys():\n",
    "            if _5_prime_nt in ['A','T'] and _3_prime_nt not in ['A','T']:\n",
    "                strand = 'TOP'\n",
    "                allele_a = ref_allele\n",
    "                allele_b = alt_allele\n",
    "                return allele_a, allele_b, strand\n",
    "            elif _3_prime_nt in ['A','T'] and _5_prime_nt not in ['A','T']:\n",
    "                strand = 'BOT'\n",
    "                allele_a = alt_allele\n",
    "                allele_b = ref_allele\n",
    "                return allele_a, allele_b, strand\n",
    "            else:\n",
    "                return seq, ref_allele, alt_allele, _5_prime_nt, _3_prime_nt\n",
    "\n",
    "def assign_ab(allele_a=None, allele_b=None, strand=None):\n",
    "    \"\"\"\n",
    "    Assign the AB genotype based on the nucleotide genotype and strand\n",
    "    Args:\n",
    "        allele_a: str, one of A, C, G, T\n",
    "        allele_b: str, one of A, C, G, T\n",
    "        strand: str, \n",
    "\n",
    "    Returns:\n",
    "        tuple (allele_a, allele_b)\n",
    "    \"\"\"\n",
    "    if allele_a is None:\n",
    "        raise ValueError(\"allele_a should not be None\")\n",
    "    if allele_b is None:\n",
    "        raise ValueError(\"allele_b should not be None\")\n",
    "    if strand is None:\n",
    "        raise ValueError(\"strand should not be None\")\n",
    "\n",
    "    genotype = allele_a + allele_b\n",
    "\n",
    "    if strand == 'TOP':\n",
    "        if genotype == 'AA':\n",
    "            return 'A','A'\n",
    "        elif genotype == 'TT':\n",
    "            return 'B','B'\n",
    "        elif genotype == 'CC':\n",
    "            return 'B','B'\n",
    "        elif genotype == 'GG':\n",
    "            return 'B','B'\n",
    "        elif genotype == 'AT':\n",
    "            return 'A','B'\n",
    "        elif genotype == 'CG':\n",
    "            return 'A','B'\n",
    "\n",
    "    elif strand == 'BOT':\n",
    "        if genotype == 'AA':\n",
    "            return 'B','B'\n",
    "        elif genotype == 'TT':\n",
    "            return 'A','A'\n",
    "        elif genotype == 'CC':\n",
    "            return 'B','B'\n",
    "        elif genotype == 'GG':\n",
    "            return 'B','B'\n",
    "        elif genotype == 'TA':\n",
    "            return 'A','B'\n",
    "        elif genotype == 'GC':\n",
    "            return 'A','B'\n",
    "\n",
    "    else:\n",
    "        raise ValueError(\"Strand should be either TOP or BOT\")\n",
    "\n",
    "\n",
    "def convert_genotype(genotype, ref_allele, alt_allele):\n",
    "    if genotype == '0/0':\n",
    "        return ref_allele, ref_allele\n",
    "    elif genotype == '1/1':\n",
    "        return alt_allele, alt_allele\n",
    "    elif genotype == '0/1':\n",
    "        return ref_allele, alt_allele\n",
    "\n",
    "    elif genotype == './.':\n",
    "        return '-', '-'\n",
    "    elif genotype == (None,None):\n",
    "        return '-', '-'\n",
    "    else:\n",
    "        raise ValueError(f\"Genotype {genotype} not recognized\")\n",
    "\n",
    "def tuple_to_float(x):\n",
    "    if x == ('A','A'):\n",
    "        return 0.\n",
    "    elif x == ('A','B'):\n",
    "        return 1.\n",
    "    elif x == ('B','B'):\n",
    "        return 2.\n",
    "    else:\n",
    "        return np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7489d96f86d40473",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:40:55.611886Z",
     "start_time": "2024-08-27T00:40:55.439784Z"
    }
   },
   "outputs": [],
   "source": [
    "REF = '../../../REFERENCES/bosTau6/bosTau6_w_wagyuY.fa'\n",
    "reference = Fasta(REF)\n",
    "SAMPLES = '/Users/callummacphillamy/Projects/bovine_pangenome/TuWa_manuscript/metadata/25_animals_MCG.xlsx'\n",
    "samples = pd.read_excel(SAMPLES)\n",
    "samples.head()\n",
    "sample_ids = samples['ID'].to_list()\n",
    "\n",
    "sample_dict = {}\n",
    "for sample in samples.itertuples():\n",
    "    if sample.No < 10:\n",
    "        sample_no = f'sample0{sample.No}'\n",
    "    else:\n",
    "        sample_no = f'sample{sample.No}'\n",
    "    \n",
    "    sample_dict[sample_no] = sample.ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fdd7fca6e681d68",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:41:00.160056Z",
     "start_time": "2024-08-27T00:41:00.037787Z"
    }
   },
   "outputs": [],
   "source": [
    "snp_map = pd.read_csv('../../../GRIP/GRIP_data/Providers/3D_Genetics/Clients/3D_Genetics/Populations/3DG_Nucleus/Standard_Analyses/we_3D_pub_03022023_Versa50K_3/we_3D_pub_snp_map_03022023_Versa50K_3.txt',\n",
    "                      sep='\\t')\n",
    "snp_map.head()\n",
    "snp_map_dict = {}\n",
    "for row in snp_map.itertuples():\n",
    "    chr_pos = f'{row[3]}_{row[4]}'\n",
    "    alleles = row.SNP.strip('[]').split('/')\n",
    "    snp_map_dict[chr_pos] = {0.:f'{alleles[0]}{alleles[0]}',\n",
    "                             1.:f'{alleles[0]}{alleles[1]}',\n",
    "                             2.:f'{alleles[1]}{alleles[1]}'}\n",
    "snp_map_dict['1_83267242']\n",
    "snp_map[snp_map['Chromosome'] == '29']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "784a83b9e06f9f82",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:41:17.776954Z",
     "start_time": "2024-08-27T00:41:17.716002Z"
    }
   },
   "outputs": [],
   "source": [
    "ARR = '../SNP_analyses/genotypes/3DG_Nucleus_array_genotypes.parquet'\n",
    "arr = pd.read_parquet(ARR)\n",
    "#arr[arr['Chr'] == '29']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8f966c61d26c434",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:41:22.350417Z",
     "start_time": "2024-08-27T00:41:22.305068Z"
    }
   },
   "outputs": [],
   "source": [
    "# Load the lifted over SNPs\n",
    "#snps = pd.read_csv('../SNP_analyses/UMD3.1.1_to_ARS-UCD1.2.versa50K.bed', sep='\\t', names=['chromosome','start','stop','UMD3.1_coords','-'])\n",
    "snps = pd.read_csv('../SNP_analyses/SNP_bed_files/versa50K_SNPs.liftover.bed', sep='\\t', names=['chromosome','start','stop'])\n",
    "#snps = snps.iloc[:, :-1]\n",
    "new_snps = snps['chromosome'].str.replace('chr','').str.cat(snps['stop'].astype(str), sep='_').to_list()\n",
    "#new_snps = snps['chromosome'].str.replace('chr','').str.cat(snps['start'].astype(str), sep='_')\n",
    "#new_snps_alt = snps['chromosome'].str.replace('chr','').str.cat(snps['stop'].astype(str), sep='_')\n",
    "# #new_snps = new_snps.to_list()\n",
    "#print(f'SNPs able to be lifted over: {len(new_snps)}')\n",
    "#print(f'SNPs able to be lifted over: {len(new_snps_alt)}')\n",
    "print(f'SNPs: ', len(new_snps))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "430a9117993da7b3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:41:09.828654Z",
     "start_time": "2024-08-27T00:41:09.823616Z"
    }
   },
   "outputs": [],
   "source": [
    "snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a53bbd706caffe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-23T06:09:45.022949Z",
     "start_time": "2024-08-23T06:09:45.020063Z"
    }
   },
   "outputs": [],
   "source": [
    "# # Convert SNPs for the Arr data to ARS-UCD1.2\n",
    "# snp_umd_to_ars_dict = {}\n",
    "# for snp in tqdm(snps.itertuples()):\n",
    "#     umd_chrom, umd_snp = re.split(r'[:-]', snp[-1])[:2]\n",
    "#     \n",
    "#     umd_chrom = umd_chrom.replace('chr','')\n",
    "#     umd_chrpos = f'{umd_chrom}_{umd_snp}'\n",
    "#     \n",
    "#     #print(umd_chrpos)\n",
    "#     \n",
    "#     ars_chrom = snp[1].replace('chr','')\n",
    "#     ars_pos = snp[2]\n",
    "#     \n",
    "#     ars_chrpos = f'{ars_chrom}_{ars_pos}'\n",
    "#     \n",
    "#     snp_umd_to_ars_dict[umd_chrpos] = ars_chrpos\n",
    "#     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d333b4aeafdcb81f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-23T06:09:45.032159Z",
     "start_time": "2024-08-23T06:09:45.030494Z"
    }
   },
   "outputs": [],
   "source": [
    "# ars_pos_col = []\n",
    "# for chr_pos in arr['ChrPos']:\n",
    "#     if chr_pos in snp_umd_to_ars_dict.keys():\n",
    "#         ars_pos_col.append(snp_umd_to_ars_dict[chr_pos])\n",
    "#     else:\n",
    "#         ars_pos_col.append('0_0')\n",
    "# arr.insert(4, 'ARS-UCD1.2_ChrPos', ars_pos_col)\n",
    "# arr_50k = arr[arr['ARS-UCD1.2_ChrPos'] != '0_0']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8114bf29ee3dbaac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:42:28.395590Z",
     "start_time": "2024-08-27T00:42:28.345082Z"
    }
   },
   "outputs": [],
   "source": [
    "#arr_50k[arr_50k['Chr'] == '29']\n",
    "arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e3010de587d1e9b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:42:35.369272Z",
     "start_time": "2024-08-27T00:42:35.286042Z"
    }
   },
   "outputs": [],
   "source": [
    "# Load the SNP map for the 50k array\n",
    "snp_map = pd.read_csv('/Users/callummacphillamy/Projects/GRIP/GRIP_data/Providers/3D_Genetics/Clients/3D_Genetics/Populations/3DG_Nucleus/Standard_Analyses/we_3D_pub_03022023_Versa50K_3/we_3D_pub_snp_map_03022023_Versa50K_3.txt', sep='\\t')\n",
    "\n",
    "snp_map_dict = {}\n",
    "for snp in snp_map.itertuples():\n",
    "    snp_map_dict[snp.Name] = snp.SNP\n",
    "snp_map_dict['BovineHD0100000033']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d28100fcf9213cc8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-23T07:29:13.288761Z",
     "start_time": "2024-08-23T07:29:13.286239Z"
    }
   },
   "outputs": [],
   "source": [
    "# Only run once\n",
    "# AGGREGATE = '../SNP_analyses/reference/3DG_Nucleus-8685_137472-Aggregated_Genotype_Counts-2024_03_14_13_33_09.833399.tsv.gz'\n",
    "# aggregate = pd.read_csv(AGGREGATE, sep='\\t', comment='#', dtype={1:str,2:str})\n",
    "# aggregate = aggregate[aggregate['ChrPos'] != '0_0']\n",
    "# aggregate = aggregate[aggregate['Chr'] != '0']\n",
    "# aggregate = aggregate[aggregate['ChrPos'].notna()]\n",
    "# aggregate = aggregate[aggregate['Pos'].notna()]\n",
    "# aggregate['Pos'] = aggregate['Pos'].astype(int)\n",
    "# \n",
    "# aggregate_info = aggregate.iloc[:, :4]\n",
    "# aggregate_samples = aggregate.loc[:, sample_ids]\n",
    "# new_aggregate = pd.concat([aggregate_info, aggregate_samples], axis='columns')\n",
    "# new_aggregate.to_parquet('../SNP_analyses/genotypes/3DG_Nucleus_array_genotypes.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f009c7d2af20930",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:42:39.985999Z",
     "start_time": "2024-08-27T00:42:39.912218Z"
    }
   },
   "outputs": [],
   "source": [
    "arr_50k = arr[arr['ChrPos'].isin(new_snps)]\n",
    "arr_50k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2977b2f0fc3b05b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T00:42:44.502785Z",
     "start_time": "2024-08-27T00:42:44.499658Z"
    }
   },
   "outputs": [],
   "source": [
    "sample_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b795a4e9b15fc6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T01:46:15.622086Z",
     "start_time": "2024-08-27T01:46:15.203301Z"
    }
   },
   "outputs": [],
   "source": [
    "ONT_SNPS = '../SNP_analyses/genotypes/mapped_to_bosTau6/sample12.versa50K.vcf.gz'\n",
    "#_50k_snps = arr_50k['ARS-UCD1.2_ChrPos'].to_list()\n",
    "\n",
    "def genotype_to_nuc_genotype(genotype, ref, alt):\n",
    "    if genotype == ('.','.'):\n",
    "        nuc_genotype = '--'\n",
    "    else:\n",
    "        genotype = tuple(map(int, genotype))\n",
    "        alleles = list(ref) + list(alt)\n",
    "        nuc_genotype = alleles[genotype[0]] + alleles[genotype[1]]\n",
    "    return nuc_genotype\n",
    "\n",
    "vcf = VariantFile(ONT_SNPS)\n",
    "for record in tqdm(vcf.fetch()):\n",
    "    samples = record.samples.keys()\n",
    "    break\n",
    "samples = list(samples)\n",
    "overlap = 0\n",
    "\n",
    "vcf = pd.read_csv(ONT_SNPS, sep='\\t', comment='#',\n",
    "                  header=None,\n",
    "                  names=['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT'] + samples)\n",
    "\n",
    "ont_df_dict = {'ChrPos':[],\n",
    "               'sample':[],\n",
    "               'genotype':[],\n",
    "               'allele_gt':[],\n",
    "               'depth':[]}\n",
    "samples = vcf.columns[9:]\n",
    "samples = [sample_dict[sample] for sample in samples]\n",
    "for row in tqdm(vcf.itertuples()):\n",
    "\n",
    "    chr_pos = str(row.CHROM).replace('chr','') + '_' + str(row.POS)\n",
    "    #if chr_pos in _50k_snps:\n",
    "    ref = row.REF\n",
    "    alt = row.ALT\n",
    "    \n",
    "    \n",
    "    locus_genotypes = list(map(lambda gt: gt.split(':')[0], row[10:]))\n",
    "    vcf_genotypes = list(map(lambda gt: tuple(gt.split('/')), locus_genotypes))     \n",
    "    allele_genotypes = list(map(lambda gt: genotype_to_nuc_genotype(gt, ref, alt), vcf_genotypes))\n",
    "    \n",
    "    if row.INFO == '.':\n",
    "        continue\n",
    "    else:\n",
    "        depth = int(row.INFO.split(';')[0].split('=')[1])\n",
    "    \n",
    "    if len(vcf_genotypes) != len(samples):\n",
    "        print('Length of genotypes does not match length of samples')\n",
    "        print(len(vcf_genotypes), len(samples))\n",
    "        print(row)\n",
    "        break\n",
    "    \n",
    "    ont_df_dict['ChrPos'].extend([chr_pos] * len(samples))\n",
    "    ont_df_dict['sample'].extend(samples)\n",
    "    ont_df_dict['genotype'].extend(vcf_genotypes)\n",
    "    ont_df_dict['allele_gt'].extend(allele_genotypes)\n",
    "    ont_df_dict['depth'].append(depth)\n",
    "x = pd.DataFrame.from_dict(ont_df_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c49fb609a1a9a696",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T01:46:28.773840Z",
     "start_time": "2024-08-27T01:46:24.904976Z"
    }
   },
   "outputs": [],
   "source": [
    "snp_map_for_conversion = snp_map[['Chromosome','Position','SNP']]\n",
    "snp_map_for_conversion = snp_map_for_conversion.copy()\n",
    "snp_map_for_conversion['SNP'] = snp_map_for_conversion['SNP'].str.strip('[]')\n",
    "snp_map_for_conversion['ChrPos'] = snp_map_for_conversion['Chromosome'].astype(str).str.cat(snp_map_for_conversion['Position'].astype(str), sep='_')\n",
    "snp_map_for_conversion['Allele_A'] = snp_map_for_conversion['SNP'].str.split('/').str[0]\n",
    "snp_map_for_conversion['Allele_B'] = snp_map_for_conversion['SNP'].str.split('/').str[1]\n",
    "snp_map_for_conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7d61647fca14fdb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T01:46:29.006074Z",
     "start_time": "2024-08-27T01:46:28.858451Z"
    }
   },
   "outputs": [],
   "source": [
    "new_x = pd.merge(x, snp_map_for_conversion[['ChrPos','Allele_A','Allele_B']], how='inner', left_on='ChrPos', right_on='ChrPos')\n",
    "new_x_dict = {'ChrPos':[],\n",
    "              'sample':[],\n",
    "              'genotype':[],\n",
    "              'allele_gt':[],\n",
    "              'allele_A':[],\n",
    "              'allele_B':[],\n",
    "              'Illumina_gt':[],\n",
    "              'depth':[]}\n",
    "\n",
    "a_dict = {'5-7':new_x_dict,\n",
    "          '8-9':new_x_dict,\n",
    "          '10+':new_x_dict}\n",
    "\n",
    "\n",
    "complementary_bases = {'A':'T',\n",
    "                       'T':'A',\n",
    "                       'C':'G',\n",
    "                       'G':'C'}\n",
    "\n",
    "for row in tqdm(new_x.itertuples()):\n",
    "    \n",
    "    _chrpos = row.ChrPos\n",
    "    _sample = row.sample\n",
    "    _genotype = row.genotype\n",
    "    _allele_gt = row.allele_gt\n",
    "    _allele_A = row.Allele_A\n",
    "    _allele_B = row.Allele_B\n",
    "    _depth = int(row.depth)\n",
    "    \n",
    "    allele_1 = row.allele_gt[0]\n",
    "    allele_2 = row.allele_gt[1]\n",
    "    \n",
    "    if allele_1 == allele_2:\n",
    "        if allele_1 == row.Allele_A:\n",
    "            _ill_gt = 0.\n",
    "            \n",
    "        elif allele_1 == row.Allele_B:\n",
    "            _ill_gt = 2.\n",
    "        elif allele_1 == complementary_bases[row.Allele_A]:\n",
    "            _ill_gt = 0.\n",
    "        elif allele_1 == complementary_bases[row.Allele_B]:\n",
    "            _ill_gt = 2.\n",
    "        else:\n",
    "           _ill_gt = np.nan\n",
    "    \n",
    "    elif allele_1 != allele_2:\n",
    "        _ill_gt = 1.\n",
    "    else:\n",
    "        print('Something went wrong')\n",
    "        break\n",
    "    \n",
    "    if _depth < 5:\n",
    "        continue\n",
    "    elif 5 <= _depth <= 7:\n",
    "        a_dict['5-7']['ChrPos'].append(_chrpos)\n",
    "        a_dict['5-7']['sample'].append(_sample)\n",
    "        a_dict['5-7']['genotype'].append(_genotype)\n",
    "        a_dict['5-7']['allele_gt'].append(_allele_gt)\n",
    "        a_dict['5-7']['allele_A'].append(_allele_A)\n",
    "        a_dict['5-7']['allele_B'].append(_allele_B)\n",
    "        a_dict['5-7']['Illumina_gt'].append(_ill_gt)\n",
    "        a_dict['5-7']['depth'].append(_depth)\n",
    "        continue\n",
    "    elif 8 <= _depth <= 9:\n",
    "        a_dict['8-9']['ChrPos'].append(_chrpos)\n",
    "        a_dict['8-9']['sample'].append(_sample)\n",
    "        a_dict['8-9']['genotype'].append(_genotype)\n",
    "        a_dict['8-9']['allele_gt'].append(_allele_gt)\n",
    "        a_dict['8-9']['allele_A'].append(_allele_A)\n",
    "        a_dict['8-9']['allele_B'].append(_allele_B)\n",
    "        a_dict['8-9']['Illumina_gt'].append(_ill_gt)\n",
    "        a_dict['8-9']['depth'].append(_depth)\n",
    "        continue\n",
    "    elif _depth >= 10:\n",
    "        a_dict['10+']['ChrPos'].append(_chrpos)\n",
    "        a_dict['10+']['sample'].append(_sample)\n",
    "        a_dict['10+']['genotype'].append(_genotype)\n",
    "        a_dict['10+']['allele_gt'].append(_allele_gt)\n",
    "        a_dict['10+']['allele_A'].append(_allele_A)\n",
    "        a_dict['10+']['allele_B'].append(_allele_B)\n",
    "        a_dict['10+']['Illumina_gt'].append(_ill_gt)\n",
    "        a_dict['10+']['depth'].append(_depth)\n",
    "        continue\n",
    "    \n",
    "a = pd.DataFrame.from_dict(a_dict['10+'])\n",
    "a = a[a['depth'] >= 10]\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c61aa7121e5f2d9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-27T01:46:37.796995Z",
     "start_time": "2024-08-27T01:46:37.725975Z"
    }
   },
   "outputs": [],
   "source": [
    "y = arr[['ChrPos', '3DWF22T0240']]\n",
    "y = y.copy()\n",
    "z = pd.merge(a[['ChrPos','Illumina_gt']], y, how='inner', left_on='ChrPos', right_on='ChrPos')\n",
    "results_dict = {'Sample':[],\n",
    "                'missing_in_ont':[],\n",
    "                'missing_in_arr':[],\n",
    "                'missing_in_both':[],\n",
    "                'correct':[],\n",
    "                'half_correct':[],\n",
    "                'incorrect':[]}\n",
    "missing_in_ont = 0\n",
    "missing_in_arr = 0\n",
    "missing_in_both = 0\n",
    "correct = 0\n",
    "half_correct = []\n",
    "incorrect = []\n",
    "comp_dict = {'AA':'TT',\n",
    "             'TT':'AA',\n",
    "             'CC':'GG',\n",
    "             'GG':'CC',\n",
    "             'AT':'TA',\n",
    "             'TA':'AT',\n",
    "             'CG':'GC',\n",
    "             'GC':'CG'}\n",
    "comparable_records = 0\n",
    "for row in z.itertuples():\n",
    "    \n",
    "    \n",
    "    if np.isnan(row.Illumina_gt) and np.isnan(row._3):\n",
    "        missing_in_both += 1\n",
    "        continue\n",
    "    elif np.isnan(row.Illumina_gt):\n",
    "        missing_in_ont += 1\n",
    "        continue\n",
    "    elif np.isnan(row._3):\n",
    "        missing_in_arr += 1\n",
    "        continue  \n",
    "    \n",
    "    if row.Illumina_gt == row._3:\n",
    "        correct += 1\n",
    "        comparable_records += 1\n",
    "    elif np.abs(row.Illumina_gt - row._3) == 2:\n",
    "        incorrect.append(row)\n",
    "        comparable_records += 1\n",
    "    elif np.abs(row.Illumina_gt - row._3) == 1:\n",
    "        half_correct.append(row)\n",
    "        comparable_records += 1\n",
    "    \n",
    "    \n",
    "print(f'Missing in both: {missing_in_both}')\n",
    "print(f'Missing in ont: {missing_in_ont}')\n",
    "print(f'Missing in arr: {missing_in_arr}')\n",
    "print(f'Correct: {correct}')\n",
    "print(f'Half correct: {len(half_correct)}')\n",
    "print(f'Incorrect: {len(incorrect)}')\n",
    "print('Comparable records:', comparable_records)\n",
    "print('Concordance rate:', (correct / comparable_records))\n",
    "# print(f'Concordance rate: {(correct + len(half_correct)) / ((correct + len(incorrect)) + len(half_correct))}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65228b87de903a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# y = arr_50k[['ChrPos','ARS-UCD1.2_ChrPos','3DWF22T0415']]\n",
    "# y = y.copy()\n",
    "# y = y[y['ARS-UCD1.2_ChrPos'].isin(x['ChrPos'])]\n",
    "# z = pd.merge(x, y, how='inner', left_on='ChrPos', right_on='ARS-UCD1.2_ChrPos')\n",
    "# missing_in_ont = 0\n",
    "# missing_in_arr = 0\n",
    "# missing_in_both = 0\n",
    "# correct = 0\n",
    "# incorrect = []\n",
    "# for row in z.itertuples():\n",
    "#     \n",
    "#     if np.isnan(row._7) and row.allele_gt == '--':\n",
    "#         missing_in_both += 1\n",
    "#         continue\n",
    "#     elif row.allele_gt == '--':\n",
    "#         missing_in_ont += 1\n",
    "#         continue\n",
    "#     elif np.isnan(row._7):\n",
    "#         missing_in_arr += 1\n",
    "#         continue  \n",
    "#     \n",
    "#     arr_nuc_gt = snp_map_dict[row.ChrPos_y][row._7]\n",
    "#     \n",
    "#     if arr_nuc_gt == row.allele_gt:\n",
    "#         correct += 1\n",
    "#     elif arr_nuc_gt != row.allele_gt:\n",
    "#         incorrect.append(row)\n",
    "# print(f'Missing in both: {missing_in_both}')\n",
    "# print(f'Missing in ont: {missing_in_ont}')\n",
    "# print(f'Missing in arr: {missing_in_arr}')\n",
    "# print(f'Correct: {correct}')\n",
    "# #print(f'Incorrect: {incorrect}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f11ad9095affa7bf",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-08T06:35:18.747906Z",
     "start_time": "2024-08-08T06:35:18.652662Z"
    }
   },
   "outputs": [],
   "source": [
    "# ONT_SNPS = '../SNP_analyses/genotypes/ont_array_comparisons/test_17_samples_29_versa50K.vcf.gz'\n",
    "# _50k_snps = arr_50k['ARS-UCD1.2_ChrPos'].to_list()\n",
    "# \n",
    "# vcf = VariantFile(ONT_SNPS)\n",
    "# for record in tqdm(vcf.fetch()):\n",
    "#     samples = record.samples.keys()\n",
    "#     break\n",
    "# samples = list(samples)\n",
    "# overlap = 0\n",
    "# overlap_1 = 0\n",
    "# overlap_2 = 0\n",
    "# \n",
    "# vcf = pd.read_csv(ONT_SNPS, sep='\\t', comment='#',\n",
    "#                   header=None,\n",
    "#                   names=['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT'] + samples)\n",
    "# #vcf['ChrPos'] = vcf['CHROM'].astype(str).str.cat(vcf['POS'].astype(str), sep='_')\n",
    "# \n",
    "# possibly_repetitive_region = []\n",
    "# indels = []\n",
    "# multiallelic = []\n",
    "# possibly_wrong_coordinate = {}\n",
    "# cannot_find_alt = {}\n",
    "# new_records = []\n",
    "# \n",
    "# \n",
    "# for row in tqdm(vcf.itertuples()):\n",
    "#     \n",
    "#     if 'INDEL' in row.INFO:\n",
    "#         indels.append(row)\n",
    "#         \n",
    "#         continue\n",
    "#     if len(row.REF) > 1:\n",
    "#         indels.append(row)\n",
    "#         continue\n",
    "#     chr_pos = \"chr\" + str(row.CHROM) + \"_\" + str(row.POS)\n",
    "#     if len(row.ALT) > 1 and row.ALT != '<NON_REF>':\n",
    "#         multiallelic.append(row)\n",
    "#         continue\n",
    "#     try:\n",
    "#         ref_allele = dbsnp_dict[chr_pos]['Ref']\n",
    "#         alt_allele = dbsnp_dict[chr_pos]['Alt']\n",
    "#     except KeyError:\n",
    "#         possibly_wrong_coordinate[chr_pos] = (row.CHROM, row.POS)\n",
    "#         continue\n",
    "# \n",
    "#     locus_genotypes = row[10:]\n",
    "#     locus_info = row[1:10]\n",
    "#     locus_genotypes = [gt.split(':')[0] for gt in locus_genotypes]\n",
    "#     # print(locus_info)\n",
    "#     # #print(locus_genotypes)\n",
    "#     # print(allele_freq)\n",
    "#     # print(locus_genotypes)\n",
    "#     # print(locus_info)\n",
    "#     # print(ref_allele)\n",
    "#     # print(alt_allele)\n",
    "#     # break\n",
    "#     sample_genotypes = list(map(lambda gt: convert_genotype(gt, ref_allele, alt_allele), locus_genotypes))\n",
    "#     \n",
    "#     if alt_allele is not None:\n",
    "#         snp = ref_allele + alt_allele\n",
    "#         if snp in unambiguous_snps.keys():\n",
    "#             allele_a, allele_b, strand = unambiguous_snps[snp]\n",
    "#             ab_genotype = list(map(lambda gt: assign_ab(gt[0], gt[1], strand), sample_genotypes))\n",
    "#             float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "#         elif snp not in unambiguous_snps.keys():\n",
    "#             try:\n",
    "#                 seq = get_sequence(reference, (str(row.CHROM), row.POS), ref_allele)\n",
    "#             except ValueError:\n",
    "#                 ref_allele = row.ALT\n",
    "#                 alt_allele = row.REF\n",
    "#                 seq = get_sequence(reference, (str(row.CHROM), row.POS), ref_allele)\n",
    "#             allele_a, allele_b, strand = resolve_ambiguous_snp(seq, ref_allele, alt_allele)\n",
    "#             ab_genotype = list(map(lambda gt: assign_ab(gt[0], gt[1], strand), sample_genotypes))\n",
    "#             float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "#             #print(row)\n",
    "#     elif alt_allele is None:\n",
    "#         raise ValueError(\"Alt allele is None\")\n",
    "#         #print(row)\n",
    "#         # allele_a, strand = resolve_ambiguous_snp_single_allele(seq, ref_allele)\n",
    "#         # ab_genotype = list(map(lambda gt: assign_ab_single_allele(gt[0], strand), sample_genotypes))\n",
    "#         # float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "#         #print(ab_genotype)\n",
    "#         break\n",
    "#     \n",
    "#     else:\n",
    "#         print('Something unforeseen happened')  \n",
    "#         break\n",
    "#     new_record = list(locus_info)\n",
    "#     new_record.extend(float_genotype)\n",
    "#     new_records.append(new_record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe44d16e293e9da2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-05T06:43:09.255845Z",
     "start_time": "2024-08-05T06:43:08.879843Z"
    }
   },
   "outputs": [],
   "source": [
    "ONT_SNPS = glob.glob('../RESULTS/batch*_results/variants/SNPs/sample*/*/merge_output.vcf.gz')\n",
    "ONT_SNPS = [ONT_SNPS[12]]\n",
    "for ont_snp in ONT_SNPS:\n",
    "    if 'MT' in ont_snp:\n",
    "        continue\n",
    "    else:\n",
    "        sample = sample_dict[ont_snp.split('/')[5]]\n",
    "        x = load_vcf(ont_snp, sample)\n",
    "        x['ChrPos'] = x['CHROM'].astype(str).str.cat(x['POS'].astype(str), sep='_')\n",
    "        # possibly_repetitive_region = []\n",
    "        # indels = []\n",
    "        # multiallelic = []\n",
    "        # possibly_wrong_coordinate = {}\n",
    "        # cannot_find_alt = {}\n",
    "        # \n",
    "        # new_records = []\n",
    "        # \n",
    "        # for row in tqdm(x.itertuples()):\n",
    "        #     if 'INDEL' in row.INFO:\n",
    "        #         indels.append(row)\n",
    "        #         continue\n",
    "        #     if len(row.REF) > 1:\n",
    "        #         indels.append(row)\n",
    "        #         continue\n",
    "        #     chr_pos = str(row.CHROM) + \"_\" + str(row.POS)\n",
    "        #     if len(row.ALT) > 1:\n",
    "        #         multiallelic.append(row)\n",
    "        #         continue\n",
    "        #     # try:\n",
    "        #         # Get the information for this SNP from the regions file\n",
    "        #         # snp_name = regions.loc[regions['ChrPos'] == chr_pos, 'Name'].values[0]\n",
    "        #     # except IndexError:\n",
    "        #     #     print('Couldn\\'t determine SNP name from manifest. Using chr_pos_ref as snp name')\n",
    "        #     #     snp_name = str(row.CHROM) + \"_\" + str(row.POS) + \"_\" + reference_name\n",
    "        #     #     possibly_wrong_coordinate[snp_name] = (row.CHROM, row.POS)\n",
    "        #     \n",
    "        #     # allele_freq = allele_frequencies.loc[\n",
    "        #     #     allele_frequencies['ChrPos'] == f'chr{chr_pos}', '{ALLELE:FREQ}'\n",
    "        #     # ].values[0]\n",
    "        #     # ref_allele, ref_freq = allele_freq.split(';')[0].split(':')\n",
    "        #     # if len(ref_allele) > 1:\n",
    "        #     #     continue\n",
    "        #     # try:\n",
    "        #     #     alt_allele, alt_freq = allele_freq.split(';')[1].split(':')\n",
    "        #     # except IndexError:\n",
    "        #     #     try:\n",
    "        #     #         alt_allele = dbsnp_dict[row.CHROM + \"_\" + str(row.POS)]['Alt']\n",
    "        #     #     except KeyError:\n",
    "        #     #         possibly_wrong_coordinate[snp_name] = (row.CHROM, row.POS)\n",
    "        #     #         cannot_find_alt[chr_pos] = None\n",
    "        #     #         continue\n",
    "        # \n",
    "        #     if len(row.ALT) > 1:\n",
    "        #         multiallelic.append(row)\n",
    "        #         continue\n",
    "        #     \n",
    "        #     ref_allele = row.REF\n",
    "        #     alt_allele = row.ALT[0]\n",
    "        #     \n",
    "        #     locus_info = row[:10]\n",
    "        #     locus_genotype = row[-2].split(':')[0]\n",
    "        #     sample_genotype = convert_genotype(locus_genotype, ref_allele, alt_allele)\n",
    "        #     #print(sample_genotype)\n",
    "        #     # locus_genotypes = [gt.split(':')[0] for gt in locus_genotypes]\n",
    "        #     # print(locus_info)\n",
    "        #     # #print(locus_genotypes)\n",
    "        #     # print(allele_freq)\n",
    "        #     # sample_genotypes = list(map(lambda gt: convert_genotype(gt, ref_allele, alt_allele), locus_genotypes))\n",
    "        #     \n",
    "        #     # print(sample_genotypes)\n",
    "        #     #print(row.CHROM, row.POS, ref_allele)\n",
    "        #     seq = get_sequence(reference, (row.CHROM, row.POS), ref_allele)\n",
    "        #     if alt_allele is not None:\n",
    "        #         snp = ref_allele + alt_allele\n",
    "        #         if snp in unambiguous_snps.keys():\n",
    "        #             allele_a, allele_b, strand = unambiguous_snps[snp]\n",
    "        #             #ab_genotype = list(map(lambda gt: assign_ab(gt[0], gt[1], strand), sample_genotypes))\n",
    "        #             ab_genotype = assign_ab(sample_genotype[0], sample_genotype[1], strand)\n",
    "        #             #float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "        #             float_genotype = tuple_to_float(ab_genotype)\n",
    "        #         elif snp not in unambiguous_snps.keys():\n",
    "        #             try:\n",
    "        #                 allele_a, allele_b, strand = resolve_ambiguous_snp(seq, ref_allele, alt_allele)\n",
    "        #                 #ab_genotype = list(map(lambda gt: assign_ab(gt[0], gt[1], strand), sample_genotypes))\n",
    "        #                 ab_genotype = assign_ab(sample_genotype[0], sample_genotype[1], strand)\n",
    "        #                 #float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "        #                 float_genotype = tuple_to_float(ab_genotype)\n",
    "        #                 #print(row)\n",
    "        #             except TypeError:\n",
    "        #                 possibly_repetitive_region.append(row)\n",
    "        #                 continue\n",
    "        #     elif alt_allele is None:\n",
    "        #         raise ValueError(\"Alt allele is None\")\n",
    "        #         #print(row)\n",
    "        #         # allele_a, strand = resolve_ambiguous_snp_single_allele(seq, ref_allele)\n",
    "        #         # ab_genotype = list(map(lambda gt: assign_ab_single_allele(gt[0], strand), sample_genotypes))\n",
    "        #         # float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "        #         #print(ab_genotype)\n",
    "        #         break\n",
    "        #     \n",
    "        #     else:\n",
    "        #         print('Something unforeseen happened')  \n",
    "        #         break\n",
    "        #     new_record = list(locus_info)\n",
    "        #     new_record.append(float_genotype)\n",
    "        #     new_records.append(new_record)\n",
    "        #     \n",
    "        #     ont_as_ab = pd.DataFrame.from_dict(new_records)\n",
    "        #     ont_as_ab = ont_as_ab.iloc[:, 1:]\n",
    "        #     ont_as_ab.columns = x.columns.to_list()[:-1]\n",
    "        #     ont_as_ab['ChrPos'] = ont_as_ab['CHROM'].astype(str).str.cat(ont_as_ab['POS'].astype(str), sep='_')\n",
    "        #     arr_sample = arr.loc[:,['ARS-UCD1.2_ChrPos', sample]]\n",
    "        #     y = pd.merge(ont_as_ab, arr_sample, how='inner', left_on='ChrPos', right_on='ARS-UCD1.2_ChrPos')\n",
    "        #     y.to_csv('../SNP_analyses/genotypes/ont_array_comparisons/{}.csv'.format(sample))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "434b789def36a48d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-05T06:50:16.629128Z",
     "start_time": "2024-08-05T06:50:16.579781Z"
    }
   },
   "outputs": [],
   "source": [
    "common_snps = pd.read_csv('../SNP_analyses/common_SNPs.bed', sep='\\t',\n",
    "                          header=0)\n",
    "common_snps['ChrPos'] = common_snps['#chrom'].astype(str).str.cat(common_snps['start_(snpPos)'].astype(str), sep='_')\n",
    "common_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "357b558d23b50b8a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-05T07:01:33.747521Z",
     "start_time": "2024-08-05T07:01:33.124921Z"
    }
   },
   "outputs": [],
   "source": [
    "possibly_repetitive_region = []\n",
    "indels = []\n",
    "multiallelic = []\n",
    "possibly_wrong_coordinate = {}\n",
    "cannot_find_alt = {}\n",
    "\n",
    "new_records = []\n",
    "\n",
    "common_snp_list = {chrpos:None for chrpos in common_snps['ChrPos']}\n",
    "common_snp_list_1 = {}\n",
    "for row in common_snps.itertuples():\n",
    "    common_snp_list_1[f\"{row[1]}_{row[3]}\"] = None\n",
    "for row in tqdm(x.itertuples()):\n",
    "    if row.ChrPos not in common_snp_list.keys():\n",
    "        continue\n",
    "    # elif row.ChrPos not in common_snp_list_1.keys():\n",
    "    #     continue\n",
    "    if 'INDEL' in row.INFO:\n",
    "        indels.append(row)\n",
    "        continue\n",
    "    if len(row.REF) > 1:\n",
    "        indels.append(row)\n",
    "        continue\n",
    "    chr_pos = str(row.CHROM) + \"_\" + str(row.POS)\n",
    "    if len(row.ALT) > 1:\n",
    "        multiallelic.append(row)\n",
    "        continue\n",
    "    # try:\n",
    "        # Get the information for this SNP from the regions file\n",
    "        # snp_name = regions.loc[regions['ChrPos'] == chr_pos, 'Name'].values[0]\n",
    "    # except IndexError:\n",
    "    #     print('Couldn\\'t determine SNP name from manifest. Using chr_pos_ref as snp name')\n",
    "    #     snp_name = str(row.CHROM) + \"_\" + str(row.POS) + \"_\" + reference_name\n",
    "    #     possibly_wrong_coordinate[snp_name] = (row.CHROM, row.POS)\n",
    "    \n",
    "    # allele_freq = allele_frequencies.loc[\n",
    "    #     allele_frequencies['ChrPos'] == f'chr{chr_pos}', '{ALLELE:FREQ}'\n",
    "    # ].values[0]\n",
    "    # ref_allele, ref_freq = allele_freq.split(';')[0].split(':')\n",
    "    # if len(ref_allele) > 1:\n",
    "    #     continue\n",
    "    # try:\n",
    "    #     alt_allele, alt_freq = allele_freq.split(';')[1].split(':')\n",
    "    # except IndexError:\n",
    "    #     try:\n",
    "    #         alt_allele = dbsnp_dict[row.CHROM + \"_\" + str(row.POS)]['Alt']\n",
    "    #     except KeyError:\n",
    "    #         possibly_wrong_coordinate[snp_name] = (row.CHROM, row.POS)\n",
    "    #         cannot_find_alt[chr_pos] = None\n",
    "    #         continue\n",
    "\n",
    "    if len(row.ALT) > 1:\n",
    "        multiallelic.append(row)\n",
    "        continue\n",
    "    \n",
    "    ref_allele = row.REF\n",
    "    alt_allele = row.ALT[0]\n",
    "    \n",
    "    locus_info = row[:10]\n",
    "    locus_genotype = row[-2].split(':')[0]\n",
    "    sample_genotype = convert_genotype(locus_genotype, ref_allele, alt_allele)\n",
    "    #print(sample_genotype)\n",
    "    # locus_genotypes = [gt.split(':')[0] for gt in locus_genotypes]\n",
    "    # print(locus_info)\n",
    "    # #print(locus_genotypes)\n",
    "    # print(allele_freq)\n",
    "    # sample_genotypes = list(map(lambda gt: convert_genotype(gt, ref_allele, alt_allele), locus_genotypes))\n",
    "    \n",
    "    # print(sample_genotypes)\n",
    "    #print(row.CHROM, row.POS, ref_allele)\n",
    "    seq = get_sequence(reference, (row.CHROM, row.POS), ref_allele)\n",
    "    if alt_allele is not None:\n",
    "        snp = ref_allele + alt_allele\n",
    "        if snp in unambiguous_snps.keys():\n",
    "            allele_a, allele_b, strand = unambiguous_snps[snp]\n",
    "            #ab_genotype = list(map(lambda gt: assign_ab(gt[0], gt[1], strand), sample_genotypes))\n",
    "            ab_genotype = assign_ab(sample_genotype[0], sample_genotype[1], strand)\n",
    "            #float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "            float_genotype = tuple_to_float(ab_genotype)\n",
    "        elif snp not in unambiguous_snps.keys():\n",
    "            try:\n",
    "                allele_a, allele_b, strand = resolve_ambiguous_snp(seq, ref_allele, alt_allele)\n",
    "                #ab_genotype = list(map(lambda gt: assign_ab(gt[0], gt[1], strand), sample_genotypes))\n",
    "                ab_genotype = assign_ab(sample_genotype[0], sample_genotype[1], strand)\n",
    "                #float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "                float_genotype = tuple_to_float(ab_genotype)\n",
    "                #print(row)\n",
    "            except TypeError:\n",
    "                possibly_repetitive_region.append(row)\n",
    "                continue\n",
    "    elif alt_allele is None:\n",
    "        raise ValueError(\"Alt allele is None\")\n",
    "        #print(row)\n",
    "        # allele_a, strand = resolve_ambiguous_snp_single_allele(seq, ref_allele)\n",
    "        # ab_genotype = list(map(lambda gt: assign_ab_single_allele(gt[0], strand), sample_genotypes))\n",
    "        # float_genotype = list(map(lambda x: tuple_to_float(x), ab_genotype))\n",
    "        #print(ab_genotype)\n",
    "        break\n",
    "    \n",
    "    else:\n",
    "        print('Something unforeseen happened')  \n",
    "        break\n",
    "    new_record = list(locus_info)\n",
    "    new_record.append(float_genotype)\n",
    "    new_records.append(new_record)\n",
    "    \n",
    "    ont_as_ab = pd.DataFrame.from_dict(new_records)\n",
    "    ont_as_ab = ont_as_ab.iloc[:, 1:]\n",
    "    ont_as_ab.columns = x.columns.to_list()[:-1]\n",
    "    ont_as_ab['ChrPos'] = ont_as_ab['CHROM'].astype(str).str.cat(ont_as_ab['POS'].astype(str), sep='_')\n",
    "    arr_sample = arr.loc[:,['ARS-UCD1.2_ChrPos', sample]]\n",
    "    y = pd.merge(ont_as_ab, arr_sample, how='inner', left_on='ChrPos', right_on='ARS-UCD1.2_ChrPos')\n",
    "    #y.to_csv('../SNP_analyses/genotypes/ont_array_comparisons/{}.csv'.format(sample))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3505488165b09a0c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-05T07:01:34.365324Z",
     "start_time": "2024-08-05T07:01:34.355983Z"
    }
   },
   "outputs": [],
   "source": [
    "ont_as_ab = pd.DataFrame.from_dict(new_records)\n",
    "ont_as_ab = ont_as_ab.iloc[:, 1:]\n",
    "ont_as_ab.columns = x.columns.to_list()[:-1]\n",
    "ont_as_ab['ChrPos'] = ont_as_ab['CHROM'].astype(str).str.cat(ont_as_ab['POS'].astype(str), sep='_')\n",
    "ont_as_ab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea971fb9d340f4a3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-05T06:52:49.461008Z",
     "start_time": "2024-08-05T06:52:49.458379Z"
    }
   },
   "outputs": [],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29e119ba88ed27bb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-05T06:17:09.114171Z",
     "start_time": "2024-08-05T06:17:09.105669Z"
    }
   },
   "outputs": [],
   "source": [
    "arr_sample = arr.loc[:,['ARS-UCD1.2_ChrPos', sample]]\n",
    "arr_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4d547aa7e265417",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-05T06:17:21.719947Z",
     "start_time": "2024-08-05T06:17:21.671996Z"
    }
   },
   "outputs": [],
   "source": [
    "y = pd.merge(ont_as_ab, arr_sample, how='inner', left_on='ChrPos', right_on='ARS-UCD1.2_ChrPos')\n",
    "y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8519797a19939df2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef35c5a85bdc28d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#`~/software/gatk-4.4.0.0/gatk GenotypeGVCFs -R ~/Projects/REFERENCES/bosTau6/bosTau6.fa -V merge_output.gvcf.gz -O test.vcf.gz -L test.bed -all-sites`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "grip2ools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
